rust/datafusion/src/optimizer/projection_push_down.rs

```diff
@@ -196,12 +95,213 @@ fn get_projected_schema(
     // create the projected schema
     let mut projected_fields: Vec<Field> = Vec::with_capacity(projection.len());
     for i in &projection {
-        projected_fields.push(table_schema.fields()[*i].clone());
+        projected_fields.push(schema.fields()[*i].clone());
     }
 
     Ok((projection, Schema::new(projected_fields)))
 }
 
+/// Recursively transverses the logical plan removing expressions and that are not needed.
+fn optimize_plan(
+    plan: &LogicalPlan,
+    required_columns: &HashSet<String>, // set of columns required up to this step
+    has_projection: bool,
+) -> Result<LogicalPlan> {
+    match plan {
+        LogicalPlan::Projection {
+            input,
+            expr,
+            schema,
+        } => {
+            // projection:
+            // * remove any expression that is not required
+            // * construct the new set of required columns
+
+            let mut new_expr = Vec::new();
+            let mut new_fields = Vec::new();
+            let mut new_required_columns = HashSet::new();
+
+            // re-write schema and expressions removing un-used columns
```

Comment:
```suggestion
            // Gather all columns needed for expressions in this Projection
```
I don't think the schema is being modified here

Reply:
Fixed